In [3]:
# CELL 0
# OBTAINING AND LOADING DATA
import pandas as pd

# Load the datasets
marketing_ads_data = pd.read_csv("marketing_ads_data.csv")
sales_transaction_data = pd.read_csv("sales_transaction_data.csv")

# Preview
print("=== Marketing Ads Data ===")
print(f"shape: {marketing_ads_data.shape}")
print(f"columns: {marketing_ads_data.columns.tolist()}")
print("\n=== Preview 10 baris pertama ===")
print(marketing_ads_data.head(10).to_string(index=False))

print("\n=== Sales Transaction Data ===")
print(f"shape: {sales_transaction_data.shape}")
print(f"columns: {sales_transaction_data.columns.tolist()}")
print("\n=== Preview 10 baris pertama ===")
print(sales_transaction_data.head(10).to_string(index=False))

=== Marketing Ads Data ===
shape: (1030, 12)
columns: ['ad_id', 'campaign_name', 'audience_type', 'product_type', 'product_name', 'variant', 'date', 'impressions', 'clicks', 'spend_idr', 'sessions', 'add_to_cart']

=== Preview 10 baris pertama ===
  ad_id        campaign_name audience_type product_type                  product_name        variant         date  impressions  clicks  spend_idr  sessions  add_to_cart
AD_0032   Awareness Campaign          cold        Serum   DermaGlow Brightening Serum   Video_Review   2023-01-06         8060     253   458896.0       216           29
AD_0110   Awareness Campaign          cold    Sunscreen DermaGlow UV Shield Sunscreen   Video_Review   01/09/2023        17314     542   108070.0       452           72
AD_0137 Retargeting Campaign          warm    Sunscreen DermaGlow UV Shield Sunscreen Static_Catalog   2023-04-13        16374     180   135435.0       146            8
AD_0089 Retargeting Campaign          warm    Sunscreen DermaGlow UV Shield 

In [4]:
# CELL 1
# SCRUBBING AND PREPARING DATA

print("===== MARKETING ADS DATA =====")

# Data Structure
print("\nStruktur Data:") 
print(marketing_ads_data.info())

# Basic Statistics
print("\nStatistik Dasar:") 
print(marketing_ads_data.describe())

# Missing Values Check
print("\nMissing Values:") 
print(marketing_ads_data.isnull().sum())

# DUplicate Check
print("\nDuplikat:") 
print(marketing_ads_data.duplicated().sum())

# Unique Values Check for Important Columns
print("\n--- Unique Values Check for Important Columns:---")
print("Variant:", marketing_ads_data['variant'].unique())
print("Campaign Name:", marketing_ads_data['campaign_name'].unique())
print("Audience Type:", marketing_ads_data['audience_type'].unique())
print("Product Type:", marketing_ads_data['product_type'].unique())
print("Product Name:", marketing_ads_data['product_name'].unique())

# Data Sample
print("Data Sample:", marketing_ads_data.sample(10))


print("\n===== SALES TRANSACTION DATA =====")

# Data Structure
print("\nStruktur Data:") 
print(sales_transaction_data.info())

# Basic Statistics
print("\nStatistik Dasar:")
print(sales_transaction_data.describe())

# Missing Values Check
print("\nMissing Values:")
print(sales_transaction_data.isnull().sum())

# DUplicate Check
print("\nDuplikat:") 
print(sales_transaction_data.duplicated().sum())

# Unique Values Check for Important Columns
print("\n--- Unique Values Check for Important Columns:---")
print("Variant:", sales_transaction_data['variant'].unique())
print("Product Type:", sales_transaction_data['product_type'].unique())
print("Product Name:", sales_transaction_data['product_name'].unique())
print("Order Status:", sales_transaction_data['order_status'].unique())

# Data Sample
print("Data Sample:", sales_transaction_data.sample(10))

===== MARKETING ADS DATA =====

Struktur Data:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1030 entries, 0 to 1029
Data columns (total 12 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   ad_id          1030 non-null   object 
 1   campaign_name  1030 non-null   object 
 2   audience_type  1030 non-null   object 
 3   product_type   1030 non-null   object 
 4   product_name   1030 non-null   object 
 5   variant        1030 non-null   object 
 6   date           1030 non-null   object 
 7   impressions    1030 non-null   int64  
 8   clicks         1030 non-null   int64  
 9   spend_idr      970 non-null    float64
 10  sessions       1030 non-null   int64  
 11  add_to_cart    1030 non-null   int64  
dtypes: float64(1), int64(4), object(7)
memory usage: 96.7+ KB
None

Statistik Dasar:
        impressions       clicks     spend_idr     sessions  add_to_cart
count   1030.000000  1030.000000  9.700000e+02  1030.000000  1030.00000

# FINDINGS: 

- **Marketing Ads 3 masalah**
    - Kolom Spend IDR = 60 nulls (5.8% dari data)
    - Kolom Date format masih object, belum datetime
    - 30 duplikat
- **Sales Transaction 3 masalah**
    - Kolom Date format masih object, belum datetime
    - Kolom Revenue IDR ada nilai negatif (min: -445.000) dan nilai (max: 469 juta) kemungkinan outlier. 
    - Kolom ad_id 92 null transaksi tanpa tracking 
    - Kolom Revenue IDR ada nilai 0.0, kemungkinan status order cancelled, perlu konfirmasi. 

In [17]:
# CELL 2 
# DATA CLEANING & TRANSFORMATION

# MARKETING ADS DATA

# 1. Drop duplicates
marketing_ads_data = marketing_ads_data.drop_duplicates()
print("Marketing Ads setelah drop duplicates:", marketing_ads_data.shape)

# 2. Standarisasi kolom Date
marketing_ads_data['date'] = pd.to_datetime(
    marketing_ads_data['date'], format='mixed', dayfirst=True
)
print("Tipe data kolom date:", marketing_ads_data['date'].dtype)

# 3. Investigasi pola null sebelum treatment 
print("\n === Null spend_idr per campaign_name: ===")
print(marketing_ads_data[marketing_ads_data['spend_idr'].isnull()]['campaign_name'].value_counts())

print("\n === Null spend_idr per audience_type: ===")
print(marketing_ads_data[marketing_ads_data['spend_idr'].isnull()]['audience_type'].value_counts())

# SALES TRANSACTIONS DATA

# 1. Standarisasi kolom Date
sales_transaction_data['date'] = pd.to_datetime(sales_transaction_data['transaction_date'], format='mixed', dayfirst=True)
print("Tipe data kolom date:", sales_transaction_data['date'].dtype)

# 2. investigasi kolom revenue_idr negatif
print("\n === revenue_idr negatif per order_status ===")
print(sales_transaction_data[sales_transaction_data['revenue_idr'] < 0]['order_status'].value_counts())
print("Jumlah baris revenue negatif:", sales_transaction_data[sales_transaction_data['revenue_idr'] < 0].shape[0])

# 3. Investigasi revenue outlier ekstrim (> 10 juta)
print("\n === Revenue ekstrim (>10 juta) ===")
print(sales_transaction_data[sales_transaction_data['revenue_idr'] > 10_000_000][['transaction_id', 'revenue_idr', 'order_status']])

# Drop kolom date karena sudah ada kolom transaction_date yang lebih spesifik
sales_transaction_data = sales_transaction_data.drop(columns=['date'])
print("\nSales Transactions columns setelah drop date:")
print(sales_transaction_data.columns.tolist())


Marketing Ads setelah drop duplicates: (1000, 12)
Tipe data kolom date: datetime64[ns]

 === Null spend_idr per campaign_name: ===
Series([], Name: count, dtype: int64)

 === Null spend_idr per audience_type: ===
Series([], Name: count, dtype: int64)
Tipe data kolom date: datetime64[ns]

 === revenue_idr negatif per order_status ===
Series([], Name: count, dtype: int64)
Jumlah baris revenue negatif: 0

 === Revenue ekstrim (>10 juta) ===
Empty DataFrame
Columns: [transaction_id, revenue_idr, order_status]
Index: []

Sales Transactions columns setelah drop date:
['transaction_id', 'ad_id', 'transaction_date', 'product_type', 'product_name', 'variant', 'revenue_idr', 'order_status']


# CELL 2 - DATA CLEANING OUTPUT

# MARKETING ADS - kolom spend_idr 56 nulls: 
Null tersebar di semua campaign (Flash Sale: 26, Awareness Campaign: 17, Retargeting Campaign: 13), mayoritas pada Cold audience (43) kemungkinan missing at random. Decision: fill null dengan median per campaign_name (lebih robust dari mean karna ada outlier di spend). 

# SALES TRANSACTION DATA - 3 Decisions 
- revenue_idr negatif semua dari Refunding (209) valid secara bisnis, tidak saya drop tapi akan saya konversi ke absolut.
- 5 baris revenue ekstrim (266 juta - 469 juta) dengan status success, asumsi merupakan data anomali dan bukan transaksi normal produk skincare, decision: drop. 
- kolom ad_id null 92 baris tidak di drop karena transaksi tetap valid untuk analisis revenue. 

In [18]:
# CELL 3
# DATA TREATMENT

# MARKETING ADS DATA

# Fill null spend_idr dengan median per campaign_name
marketing_ads_data['spend_idr'] = marketing_ads_data.groupby('campaign_name')['spend_idr'].transform(lambda x: x.fillna(x.median()))
print("Null spend_idr setelah fill:", marketing_ads_data['spend_idr'].isnull().sum())

# SALES TRANSACTIONS DATA

# 1. Drop revenue ekstrim 
before = sales_transaction_data.shape[0]
sales_transaction_data = sales_transaction_data[sales_transaction_data['revenue_idr'] <= 10_000_000]
print(f"Baris di drop karna revenue ekstrim: {before - sales_transaction_data.shape[0]}")

# 2. Revenue negatif Refunding di konversi ke absolut
sales_transaction_data['revenue_idr'] = sales_transaction_data['revenue_idr'].abs()
print("Min revenue_idr setelah konversi:", sales_transaction_data['revenue_idr'].min())

# FINAL CHECK

print("\n=== Final Shape ===")
print("Marketing Ads:", marketing_ads_data.shape)
print("Sales Transactions:", sales_transaction_data.shape)

print("\n=== Final Null Check ===")
print(marketing_ads_data.isnull().sum())
print(sales_transaction_data.isnull().sum())


Null spend_idr setelah fill: 0
Baris di drop karna revenue ekstrim: 0
Min revenue_idr setelah konversi: 0.0

=== Final Shape ===
Marketing Ads: (1000, 12)
Sales Transactions: (2495, 8)

=== Final Null Check ===
ad_id            0
campaign_name    0
audience_type    0
product_type     0
product_name     0
variant          0
date             0
impressions      0
clicks           0
spend_idr        0
sessions         0
add_to_cart      0
dtype: int64
transaction_id       0
ad_id               92
transaction_date     0
product_type         0
product_name         0
variant              0
revenue_idr          0
order_status         0
dtype: int64


# CELL 3 
# SUMMARY

1. kolom spend_idr null diisi dengan nilai median per campaign tidak ada lagi kolom null tersisa.
2. revenue ekstrim sudah tidak ada. 
3. revenue negatif di konversi ke absolut.
4. Shape Final: Marketing Ads: (1000, 12), Sales Transactions: (2495, 8).